### PR TITLE
Feature/issue 3 cache historical prices

### PR DESCRIPTION
**Pull request by chasedmorell**

### CODE REVIEW

File: src/main/java/org/galatea/starter/domain/IexHistoricalPrices.java [hunk 1]
```diff
@@ -1,19 +1,31 @@
...
+import org.galatea.starter.domain.rpsy.IexHistoricalPricesKey;
 
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.IdClass;
+import javax.xml.bind.annotation.XmlRootElement;
 import java.math.BigDecimal;
 import java.util.Date;
 
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/domain/IexHistoricalPrices.java:15]: Why did you need to set this explicitly? Should the @Builder have given you an AllArgsConstructor anyway?

File: src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package org.galatea.starter.domain.rpsy;
+
+import java.io.Serializable;
+import java.util.Date;
+
+public class IexHistoricalPricesKey implements Serializable {
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java:18]: Does this need Serializable on it? The cache entry isn't leaving the JVM right?. You can also use Lombok instead of the explicit constructors. I'd recommend just using @Data because then you also get @ToString and @EqualsAndHashcode for free

File: src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesRpsy.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  /**
+   * Retrieves all entities with the given symbol.
+   */
+  List<IexHistoricalPrices> findBySymbol(String symbol);
+
+  @Override
+  List<IexHistoricalPrices> findAllById(Iterable<IexHistoricalPricesKey> ids);
+
+  /**
+   * 'p0' required in key because java does not retain parameter names during compilation unless
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesRpsy.java]: What is this comment about?

File: src/main/java/org/galatea/starter/utils/Helpers.java [hunk 4]
```diff
@@ -41,4 +52,26 @@ public static DiffResult diff(final Object lhs, final Object rhs, final MethodFi
 
     return builder.build();
   }
+
+  public static List<Date> getBusinessDaysSinceDate(final LocalDate endDate, final LocalDate startDate) {
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/utils/Helpers.java]: Reasonable approach... but wouldn't actually work in practice. The reason is bank holidays -- e.g. Christmas may not fall on a weekend, but would still not have any data for it. This method wouldn't exclude it, and then we'd unnecessarily be hitting the API. Date logic is the worst kind of logic... Let me know when you see this and we can discuss alternative strategies for caching the data.

File: src/main/java/org/galatea/starter/entrypoint/IexRestController.java [hunk 5]
```diff
@@ -53,13 +53,14 @@ public List<IexLastTradedPrice> getLastTradedPrice(
...
-   * @param range the range of the historical data, starting at today. (examples: 1m, 5d)
+   * @param range the range of the historical data, starting at today. Unit is days.
+   *              For example, range=5 will return the last 5 days of prices, assuming market was open for all 5 days.
    * @return a List of IexHistoricalPrices objects for the given symbol and range.
    */
   @GetMapping(value = "${mvc.iex.getHistoricalPricesPath}", produces = {
           MediaType.APPLICATION_JSON_VALUE})
   public List<IexHistoricalPrices> getHistoricalPricesForSymbol(
-          @RequestParam(value = "symbol") final String symbol, @RequestParam(value="range") final String range) {
+          @RequestParam(value = "symbol") final String symbol, @RequestParam(value="range") final Integer range) {
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/entrypoint/IexRestController.java]: Questionable decision here -- I'd assume the IEX API allows for different "range" formats -- e.g. d for days, m for months, etc. You've basically forced the controller to only allow "day" queries. So, reduced functionality in order to make the caching a little easier to reason about. Let's discuss.

If the IEX API only does days, then that's a different matter.

File: src/main/java/org/galatea/starter/domain/IexHistoricalPrices.java (hunk 1 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/domain/IexHistoricalPrices.java:15]: I believe I need to explicitly set this because I've declared a constructor in the class (it creates a historical price object from a key object). Without the @AllArgsConstructor, I get no valid constructor found errors.

File: src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java (hunk 2 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java:18]: Because this is being used as a composite key, it required the key to implement serializable. I'm not sure why, internally, this is the case. 

I'll use @Data instead of explicit constructors.

File: src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesRpsy.java (hunk 3 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesRpsy.java]: I have no clue, will remove.

File: src/main/java/org/galatea/starter/utils/Helpers.java (hunk 4 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/utils/Helpers.java]: Cache strategy changed.

File: src/main/java/org/galatea/starter/entrypoint/IexRestController.java (hunk 5 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/entrypoint/IexRestController.java]: Changed to allow for days, months, year & quarters.

File: src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    public String symbol;
+    public Date date;
+
+    public IexHistoricalPricesKey(String symbol, Date date){
+        this.symbol = symbol;
+        this.date = date;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java]: You shouldn't need to write these methods once you put @Data on it

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 7]
```diff
@@ -55,7 +64,77 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
    * @return a list of IexHistoricalPrices objects
    */
   public List<IexHistoricalPrices> getHistoricalPricesForSymbol(final String symbol, final String range) {
-      return iexClient.getHistoricalPricesForSymbol(symbol, range);
+
+      LocalDate sinceDate;
+      Integer rangeValue = Integer.parseInt(range.replaceAll("[^0-9]",""));
+
+      switch(range.charAt(range.length()-1)){
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/service/IexService.java]: This is a reasonable use of a switch-case statement, but we try and stay away from them because they lead to violations of the Open/Closed principle. Try using enums here following the Strategy pattern (similar to what we were going to cover in yesterday's book club).

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 8]
```diff
@@ -55,7 +64,77 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
...
+          case 'd': sinceDate = LocalDate.now().minusDays(rangeValue); break;
+          case 'm': sinceDate = LocalDate.now().minusMonths(rangeValue); break;
+          case 'y': sinceDate = LocalDate.now().minusYears(rangeValue); break;
+          case 'q': sinceDate = LocalDate.now().minusMonths(rangeValue*3); break;
+          default: return new ArrayList<IexHistoricalPrices>();
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/service/IexService.java]: Is this really what you want to do here? This case would hit if a caller of this class used any invalid period specifier -- returning an empty list is misleading because the caller would reasonably interpret that as meaning there's no data for their request. In reality, their request was bad -- so this is a place we should throw an Exception. Java's built in IllegalArgumentException works fine here, no need to define a custom exception type.

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 9]
```diff
@@ -55,7 +64,77 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
...
+      //when a record isn't found in cache, immediately stop querying cache.
+      List<IexHistoricalPrices> cachedHistoricalPrices = new ArrayList<>();
+      Boolean allDatesFoundInCache = true;
+
+      for(IexHistoricalPricesKey key: keys){
+          Optional<IexHistoricalPrices> price = iexHistoricalPricesRpsy.findById(key);
+          //Price records for days that market is closed (weekends, etc) have null values.
+          //Don't include these in the response.
+          if(price.isPresent()){
+              if((price.get().getClose().isPresent())){
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/service/IexService.java]: Nit - Seems like you got an extra pair of parens here

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 10]
```diff
@@ -55,7 +64,77 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
...
+      }
+
+      //find all the historicalPrice records, if they exist, in cache database.
+      //when a record isn't found in cache, immediately stop querying cache.
+      List<IexHistoricalPrices> cachedHistoricalPrices = new ArrayList<>();
+      Boolean allDatesFoundInCache = true;
+
+      for(IexHistoricalPricesKey key: keys){
+          Optional<IexHistoricalPrices> price = iexHistoricalPricesRpsy.findById(key);
+          //Price records for days that market is closed (weekends, etc) have null values.
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/service/IexService.java]: This comment doesn't exactly line up with the code under -- when I read "have null values" I would expect that means if the Optional.isPresent() method returns false, that means it may be a weekend - but the code implies that optional.get().getClose().isPresent() is where the weekend case is dealt with.

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 11]
```diff
@@ -55,7 +64,77 @@ public List<IexLastTradedPrice> getLastTradedPriceForSymbols(final List<String>
    * @return a list of IexHistoricalPrices objects
    */
   public List<IexHistoricalPrices> getHistoricalPricesForSymbol(final String symbol, final String range) {
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/service/IexService.java:65]: This method is getting long... I'd recommend splitting it up into smaller methods.

File: src/main/java/org/galatea/starter/utils/Helpers.java [hunk 12]
```diff
@@ -41,4 +52,22 @@ public static DiffResult diff(final Object lhs, final Object rhs, final MethodFi
...
   }
+
+  public static List<Date> getDaysSinceDate(final LocalDate endDate, final LocalDate startDate) {
+
+
+    // Iterate over stream of all dates
+    List<LocalDate> days = startDate.datesUntil(endDate)
+            .collect(Collectors.toList());
+
+    List<Date> DaysDate = new ArrayList<Date>();
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/utils/Helpers.java]: "daysDate" rather than "DaysDate"

File: src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java (hunk 6 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java]: Done.

File: src/main/java/org/galatea/starter/service/IexService.java (hunk 7 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/service/IexService.java]: Thank you, this is a good idea. Will do.

File: src/main/java/org/galatea/starter/service/IexService.java (hunk 8 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/service/IexService.java]: Done.

File: src/main/java/org/galatea/starter/service/IexService.java (hunk 9 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/service/IexService.java]: You're right, good call.

File: src/main/java/org/galatea/starter/service/IexService.java (hunk 10 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/service/IexService.java]: Fixed comment.

File: src/main/java/org/galatea/starter/utils/Helpers.java (hunk 12 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/utils/Helpers.java]: Done.

File: src/main/java/org/galatea/starter/domain/RangeType.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
+package org.galatea.starter.domain;
+
+import java.time.LocalDate;
+import java.util.Optional;
+
+public enum RangeType {
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/domain/RangeType.java:7]: This looks good! You can take it one step further though so that you can just specify a lambda rather than having to write the full method each time.

Make the subtract() method final and change its type to Function<Integer, LocalDate>. It doesn't need to be abstract any more either. Add it into the constructor.

Then, in the enum constructors, you can do:
Days("d", rangeValue -> LocalDate.now().minusDays(rangeValue)).

Thus setting the subtract() method to the lambda you pass in.

File: src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java [hunk 14]
```diff
@@ -1,36 +1,52 @@
...
+        Optional<RangeType> type = RangeType.fromString(String.valueOf(range.charAt(range.length()-1)));
 
-    @Override
-    public boolean equals(Object obj) {
-        if(!(obj.getClass().equals(this.getClass()))){
-            return false;
+        if(type.isPresent()){
+            sinceDate = type.get().subtract(rangeValue);
+        }else{
+            throw new IllegalArgumentException();
```
Comment (Hurryitup) [inline src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java]: You should put an explanation of what caused the exception into the Exception itself -- an empty Exception wouldn't be helpful if you were to be woken up at 5am to a call from support :) 

`throw new IllegalArgumentException(String.format("Invalid range type specified: %s", type))`

You should put the value from the String.valueOf(range....) call into a variable so that you can log it out as above.

Comment (chasedmorell) [reply src/main/java/org/galatea/starter/domain/rpsy/IexHistoricalPricesKey.java]: Good point!

File: src/main/java/org/galatea/starter/domain/RangeType.java (hunk 13 above)
Comment (chasedmorell) [reply src/main/java/org/galatea/starter/domain/RangeType.java:7]: Done.

**Review by Hurryitup — APPROVED** [review]
